Realm/RLMLogger.mm

```diff
@@ -0,0 +1,171 @@
+        case Level::all:    return @"";
+        case Level::trace:  return @"Trace";
+        case Level::debug:  return @"Debug";
+        case Level::detail: return @"Detail";
+        case Level::info:   return @"Info";
+        case Level::error:  return @"Error";
+        case Level::warn:   return @"Warning";
+        case Level::fatal:  return @"Fatal";
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+struct CocoaLogger : public Logger {
+    void do_log(Level level, const std::string& message) override {
+        NSLog(@"%@: %@", levelPrefix(level), RLMStringDataToNSString(message));
+    }
+};
+
+class CustomLogger : public Logger {
+public:
+    RLMLoggerFunction function;
+    void do_log(Level level, const std::string& message) override {
+        @autoreleasepool {
+            if (function) {
+                function(logLevelForLevel(level), RLMStringDataToNSString(message));
+            }
+        }
+    }
+};
+} // anonymous namespace
+
+@implementation RLMLogger {
+    std::shared_ptr<Logger> _logger;
+}
+
+static RLMLogger *s_defaultLogger;
+
+typedef void(^LoggerBlock)(RLMLogLevel level, NSString *message);
+
+- (RLMLogLevel)level {
+    return logLevelForLevel(Logger::get_default_level_threshold());
+}
+
+- (void)setLevel:(RLMLogLevel)level {
+    Logger::set_default_level_threshold(levelForLogLevel(level));
+}
+
+- (instancetype)initWithLogger:(std::shared_ptr<Logger>)logger {
+    if (self = [self init]) {
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (instancetype)initWithLevel:(RLMLogLevel)level logFunction:(RLMLogFunction)logFunction {
+    if (self = [super init]) {
+        auto logger = std::make_shared<CustomLogger>();
+        logger->set_level_threshold(levelForLogLevel(level));
+        logger->function = logFunction;
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (void)logWithLevel:(RLMLogLevel)logLevel message:(NSString *)message, ... {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        va_list args;
+        va_start(args, message);
+        _logger->log(level, message.UTF8String, args);
+        va_end(args);
+    }
+}
+
+- (void)logLevel:(RLMLogLevel)logLevel message:(NSString *)message {
+    auto level = levelForLogLevel(logLevel);
+    if (_logger->would_log(level)) {
+        _logger->log(level, message.UTF8String);
+    }
+}
+
+#pragma mark Global Logger Setter
+
++ (instancetype)defaultLogger {
+    RLMLogger *logger;
+    if (!s_defaultLogger) {
+        auto defaultLogger = std::make_shared<CocoaLogger>();
+        defaultLogger->set_level_threshold(Level::warn);
+        s_defaultLogger = [[RLMLogger alloc] initWithLogger:defaultLogger];
+    }
+
+    logger = s_defaultLogger;
+    Logger::set_default_logger(logger->_logger);
+    return logger;
+}
+
++ (void)setDefaultLogger:(RLMLogger *)logger {
+    Logger::set_default_level_threshold(logger->_logger->get_level_threshold());
```

Comment:
No need to call `set_default_level_threshold()`, as it'll be replaced by the log level of the logger instance.